[tensorexpr] Add an optional argument for set of unallocated intermediate buffers in prepareForCodegen/insertAllocFree

huiguoo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64227
* __->__ #64226



Differential Revision: [D30652221](https://our.internmc.facebook.com/intern/diff/D30652221)

huiguoo committed:
[tensorexpr] Add 'is_allocated' flag for buffers and use it to insert 'Alloc/Free' stmts

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30652221).

ZolotukhinM reviewed (commented):
I'm not sure about this change. I believe we might be leaking an abstraction from TensorExprKernel to Buf here, and it can be avoided here without much issues. Instead of adding a flag to every Buf, TEK could simply maintain a set of allocated buffers and only insert Alloc-Free for non-allocated or even pre-allocate them at the same time it tries to insert Alloc-Free (i.e. for each buffer: try to preallocate and if fails insert alloc-free).

huiguoo commented:
I attempted to reuse `LoopNest.prepareForCodegen`  from the workflow of `construct LoopNest`+`LoopNest.prepareForCodegen` + `construct Codegen` which way users use NNC without torchscript. Currently, we insert `Alloc/Free` stmts in `prepareForCodegen` as well as expand reductions and flatten indices. We may need to move all these work to TEK if we want to separate the `Alloc/Free` insertions from it.

ZolotukhinM commented:
We're adding pre-allocated buffers to constants, which is a property of TEK, so this logic belongs there too. Unless we want to move constants out of TEK as well, I don't think we need to support this transformation (preallocation for intermediate buffers) in a separation from TEK - and I don't think there is much value in such separation?

huiguoo commented:
Oh, I mean to separate `Alloc/Free` insertions from `prepareForCodegen` if you think we should preallocate intermediate buffers at the same time we try insert `Alloc/Free` in TEK. 

In summary, we have two options if we do not use `is_allocated` flag in buf:
1) copy all work from `prepareForCodegen` to TEK, so TEK can customize the intermediate buffer allocations without affecting `prepareForCodegen`
2) separate `Alloc/Free` insertions from `prepareForCodegen`: TEK calls `prepareForCodegen` after intermediate buffer allocations; and users need to call `prepareForCodegen` + `insertAllocFree` when using NNC without torchscript (i.e., without TEK)

ZolotukhinM commented:
How about we change `insertAllocFree` to accept an optional argument `std::vector<ConstantDescr>& constants_`? If this argument is provided, then the pass tries to pre-allocate intermediate buffers when possible and `constants_` get updated. If not, or when a buffer can't be pre-allocated, Alloc-Free is inserted (the current behavior). This way we would be able to use it from TEK and without it if we need.

bertmaher reviewed (changes requested):
I could be convinced here, but I think it is a compiler anti-pattern to have random flags on IR that need to be set for downstream passes to behave correctly.  If we need to be able to insert manual allocations for intermediate buffers, I'd recommend having the `insertAllocFree` pass first look for `Alloc` statements and remove them from the set of `intermediateBufs`.  That way all the logic is in one place and you don't have to track down where or why the `is_allocated` flag is set.

bertmaher commented:
Sigh, I didn't see @ZolotukhinM's comments before reviewing :-p.  I think the proposed design satisfies my concerns too :-)

huiguoo commented:
I tend to like @bertmaher 's approach. Manual allocations may happen beyond TEK and it's more neat if we just look for `Alloc` stmts in IR to exclude those buffers in `prepareForCodegen`.

huiguoo commented:
Oh but we still need TEK to specify what buffers it has allocated at compile time (which does not insert `Alloc` stmts). So for the problem we try to solve here, I'll implement what @ZolotukhinM proposed, and add a filter to `prepareForCodegen` in another PR.

huiguoo committed:
Update on "[tensorexpr] Add an optional argument for set of unallocated intermediate buffers in prepareForCodegen/insertAllocFree"




Differential Revision: [D30652221](https://our.internmc.facebook.com/intern/diff/D30652221)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30652221).

huiguoo committed:
Update on "[tensorexpr] Add an optional argument for set of unallocated intermediate buffers in prepareForCodegen/insertAllocFree"




Differential Revision: [D30652221](https://our.internmc.facebook.com/intern/diff/D30652221)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30652221).

huiguoo committed:
Update on "[tensorexpr] Add an optional argument for set of unallocated intermediate buffers in prepareForCodegen/insertAllocFree"




Differential Revision: [D30652221](https://our.internmc.facebook.com/intern/diff/D30652221)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30652221).

huiguoo committed:
Update on "[tensorexpr] Add an optional argument for set of unallocated intermediate buffers in prepareForCodegen/insertAllocFree"




Differential Revision: [D30652221](https://our.internmc.facebook.com/intern/diff/D30652221)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30652221).

ZolotukhinM commented on torch/csrc/jit/tensorexpr/loopnest.cpp:
```diff
@@ -1024,15 +1033,17 @@ void LoopNest::eliminateDeadStores() {
 }
 
-void LoopNest::prepareForCodegen() {
+void LoopNest::prepareForCodegen(
+    const c10::optional<std::unordered_set<BufPtr>>&
+        interm_bufs /*= c10::nullopt*/) {
```
This signature is now not obvious. From external perspective it could be unclear why we need to pass a set of intermediate buffers to this function and it could be surprising that it actually controls whether we preallocate them or not. I think a cleaner interface would take a bool flag (insert alloc/free stmts or not, or in other words assume intermediate buffers are already preallocated or not).

huiguoo replied:
Is it right to assume all intermediate buffers can be preallocated, i.e., all buffers are static? Because for dynamic buffers, we can only allocate them at runtime.

Or, we can re-check the shapes of each intermediate buffer in `insertAllocFree` and always allocate buffers if dynamic; and use the bool flag to specify static buffers.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/loopnest.cpp:
```diff
@@ -59,7 +59,7 @@ LoopNest::LoopNest(const std::vector<Tensor>& output_tensors) {
 }
 
-const std::unordered_set<BufPtr> LoopNest::getIntermediateBufs() const {
+std::unordered_set<BufPtr> LoopNest::getIntermediateBufs() const {
```
I was at first surprised why we need this change but then found the answer in your second PR: we want to return mutable set so that `preallocate` function can erase preallocated bufs from it (is my understanding correct on this?). I wonder if a cleaner alternative would be to exclude buffers from `constants_` in this function (on line 67) - and to make this function stateless we can pass `constants_` to it as an optional parameter too. What do you think?

huiguoo replied:
> we want to return mutable set so that preallocate function can erase preallocated bufs from it (is my understanding correct on this?)

It's correct. This function collects all intermediate buffers into a set but the set should not be const so the user can add/delete buffers in the set. TEK modified the set (erased the buffers it pre-allocated) and passed it to `prepareForCodegen`. 

>  I wonder if a cleaner alternative would be to exclude buffers from constants_ in this function (on line 67) - and to make this function stateless we can pass constants_ to it as an optional parameter too
I am not sure if I understand you correctly. I don't think we can add filter (`constants_`) to this function from LoopNest. There are 2 call sites to this function: 1) TEK calls it to obtain the complete set of intermediate buffers and it does not need to exclude any buffers; 2) LoopNest::insertAllocFree calls it to obtain the set of intermediate buffers for runtime allocation and we need to exclude the buffers that have been preallocated in TEK (`constants_`) but how can we pass this info (`constants_`) to LoopNest from TEK?

ZolotukhinM replied:
Right, passing `constants_` to LoopNest would be annoying.

How about the following crazy idea: what if we move logic of `prepareForCodegen` to the CodeGen constructor? In CodeGen we have access to BufferArgs, which essentially include inputs+outputs+constants, so it's trivial to determine which buffers need Alloc/Free. As a side benefit, we'll also move that logic to where it belongs better: to the codegen, and then we'll have less stateful methods in LoopNest, so hopefully we could completely get rid of it. What do you think?

It might be a sizable change, so if you'd like to land this version first I'd be ok with it, but then we'll have to at least add comments explaining why and what we pass, because currently the APIs become non-obvious.

huiguoo replied:
Yeah, currently intermediate buffer allocation (`insertAllocFree`) happens implicitly in LoopNest::prepareForCodegen and if we want to allow users to specify pre-allocation optimization in TEK we'll have to either 1) make `insertAllocFree` explicit so it's available to TEK, or 2) as you suggested, move `insertAllocFree` to codegen which happens after TEK.

I kinda like approach 2) at this point. Approach 1) requires some refactoring in LoopNest and it's not clear to me how to do it yet; approach 2) seems more neat since we only need to move `insertAllocFree` from LoopNest::prepareForCodegen to codegen constructor. 

And yeah, I'd love to explore approach 2) in another PR.  It seems to be an isolated task. I'll add comments for the intermediate buffer argument in this PR :)

ZolotukhinM replied:
The PR looks good to me then! cc @bertmaher @navahgar for their opinions on (2)

huiguoo committed:
Update on "[tensorexpr] Add an optional argument for set of unallocated intermediate buffers in prepareForCodegen/insertAllocFree"




Differential Revision: [D30652221](https://our.internmc.facebook.com/intern/diff/D30652221)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30652221).

huiguoo committed:
Update on "[tensorexpr] Add an optional argument for set of unallocated intermediate buffers in prepareForCodegen/insertAllocFree"




Differential Revision: [D30652221](https://our.internmc.facebook.com/intern/diff/D30652221)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30652221).

ZolotukhinM commented:
@huiguoo, please make sure to update commit message after substantial changes to the PR. Not a big deal, but it helps to have accurate descriptions when reading git-logs. Note that you need to update phabricator diff as well (use `ghimport -u ...` to do that).

huiguoo commented:
@ZolotukhinM The git-logs have been updated since the first commit: this PR adds an optional argument in `PrepareForCodegen`/`insertAllocFree`? First commit of this PR adds a `is_alloc` field in `Buf`, lol

ZolotukhinM commented:
@huiguoo, yes, I see that you updated the PR. But I think the corresponding diff was not updated and when the PR was merged into the master it got the old commit message because of that. If you `git log` on a recent master and look for this commit, you'll see this: https://github.com/pytorch/pytorch/commit/527348a6fe934dcfeeb88352e0cb9e263c76b78d
